lnbits/wallets/nwc.py

```diff
@@ -0,0 +1,934 @@
+            if "lookup_invoice" in info["supported_methods"]:
+                created_at = int(resp.get("created_at", time.time()))
+                expires_at = int(resp.get("expires_at", created_at + 3600))
+                self.pending_payments.append(
+                    {  # Start tracking
+                        "checking_id": checking_id,
+                        "expires_at": expires_at,
+                        "settled": False,
+                        "expired": False,
+                    }
+                )
+            return InvoiceResponse(True, checking_id, payment_request, None)
+        except Exception as e:
+            return InvoiceResponse(ok=False, error_message=str(e))
+
+    async def status(self) -> StatusResponse:
+        try:
+            info = await self.conn.get_info()
+            if "get_balance" not in info["supported_methods"]:
+                logger.debug("get_balance is not supported by this NWC service.")
+                return StatusResponse(None, 0)
+            resp = await self.conn.call("get_balance", {})
+            balance = int(resp["balance"])
+            return StatusResponse(None, balance)
+        except Exception as e:
+            return StatusResponse(str(e), 0)
+
+    async def pay_invoice(self, bolt11: str, fee_limit_msat: int) -> PaymentResponse:
+        try:
+            resp = await self.conn.call("pay_invoice", {"invoice": bolt11})
+            preimage = resp.get("preimage", None)
+            invoice_data = bolt11_decode(bolt11)
+            payment_hash = invoice_data.payment_hash
+            # pay_invoice doesn't return payment data, so we need
+            # to call lookup_invoice too (if supported)
+            info = await self.conn.get_info()
+
+            if "lookup_invoice" not in info["supported_methods"]:
+                # if not supported, we assume it succeeded
+                return PaymentResponse(True, payment_hash, None, preimage, None)
+
+            try:
+                payment_data = await self.conn.call(
+                    "lookup_invoice", {"invoice": bolt11}
+                )
+                settled = payment_data.get("settled_at", None) and payment_data.get(
+                    "preimage", None
+                )
+                if not settled:
+                    return PaymentResponse(None, payment_hash, None, None, None)
+                else:
+                    fee_msat = payment_data.get("fees_paid", None)
+                    return PaymentResponse(True, payment_hash, fee_msat, preimage, None)
+            except Exception:
+                # Workaround: some nwc service providers might not store the invoice
+                # right away, so this call may raise an exception.
+                # We will assume the payment is pending anyway
+                return PaymentResponse(None, payment_hash, None, None, None)
+
+        except Exception as e:
+            return PaymentResponse(ok=False, error_message=str(e))
```

Comment:
`PaymentResponse` should failed (`ok=False`) only if the funding source said so explicitly.
If its possible that a network failure occurred (websocket closed for example) but the payment went trough then the payment should be marked as pending.

Failed payment are not subtracted from the wallet balance, so this is a double spend vulnerability.

Reply:
I added a catch for NWC specific error, that check for known failure codes, so this should solve the issue